### PR TITLE
Provide bucket replication latency and throughput prometheus statistics

### PR DESCRIPTION
**Pull request by krishnasrinivas**
## Description
Provide bucket replication latency and throughput prometheus statistics

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

### CODE REVIEW

File: cmd/metrics.go [hunk 1]
```diff
@@ -38,6 +40,13 @@ var (
 		},
 		[]string{"api"},
 	)
+	httpReplicationDuration = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "s3_ttfb_seconds",
```
Comment (klauspost) [inline cmd/metrics.go]: Why is it called `ttfb`? It doesn't seem to contain time to first byte.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally
Error: No such image: minio/minio:10358-c9b02ba

File: cmd/metrics.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/metrics.go]: It's actually s3_duration_seconds

Comment (krishnasrinivas) [reply cmd/metrics.go]: oops copy/paste problem

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-zoned.sh | :x: [more...](#6c384988-6293-457d-a130-fff6078c99ab) |
<a name="6c384988-6293-457d-a130-fff6078c99ab"></a>
### 10358-73b71ef/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:32090
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7b16f9e33ba7:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 3 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 1 minutes and 19 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 42 seconds
(8/15) Running minio-dotnet tests ... done in 47 seconds
(9/15) Running minio-go tests ... done in 2 minutes and 16 seconds
(10/15) Running minio-java tests ... done in 1 minutes and 47 seconds
(11/15) Running minio-js tests ... done in 51 seconds
(12/15) Running minio-py tests ... done in 3 minutes and 44 seconds
(13/15) Running s3cmd tests ... FAILED in 21 seconds
{
  "name": "s3cmd",
  "duration": "10139",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "WARNING: Bucket is not empty. Removing all the objects from it first. This may take some time...\nERROR: S3 error: 404 (Not Found)"
}
(13/15) Running s3select tests ... done in 8 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally
Error: No such image: minio/minio:10358-73b71ef

File: cmd/metrics.go [hunk 2]
```diff
@@ -38,6 +40,13 @@ var (
 		},
 		[]string{"api"},
 	)
+	httpReplicationDuration = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "replication_duration_seconds",
+			Help:    "Time taken by requests served by current MinIO server instance",
```
Comment (harshavardhana) [inline cmd/metrics.go]:
```suggestion
			Help:    "Time taken by replication requests served by current MinIO server instance",
```

File: cmd/metrics.go [hunk 3]
```diff
@@ -367,6 +376,15 @@ func networkMetricsPrometheus(ch chan<- prometheus.Metric) {
 		prometheus.CounterValue,
 		float64(connStats.S3InputBytes),
 	)
+
+	ch <- prometheus.MustNewConstMetric(
+		prometheus.NewDesc(
+			prometheus.BuildFQName("replication", "tx", "bytes_total"),
+			"Total number of bytes transferred due to bucket replication by current MinIO server instance",
```
Comment (harshavardhana) [inline cmd/metrics.go]:
```suggestion
			"Total number of bytes transferred in bucket replication by current MinIO server instance",
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Can we think about this from the point of view to keep these metrics per bucket level? instead of the total level? 

Keeping this per bucket level makes it easier to understand which replication is using how much? 

I suspect that bucket level granularity helps? and we can add the following details as well 

- Number of objects replicated per bucket
- Total time taken by one replication cycle to complete per bucket
- Total bytes transferred by one replication cycle per bucket

Something along the lines of but not entirely the same thing
https://aws.amazon.com/blogs/aws/s3-replication-update-replication-sla-metrics-and-events/

**Comment by krishnasrinivas** [comment]
> Can we think about this from the point of view to keep these metrics per bucket level? instead of the total level?

@kerneltime is working on a fine grained replication metrics implementation: https://github.com/miniohq/product/blob/6347113e47c11c8fb63246dac23fc5ae5708e1b5/proposals/replication.bandwidth.md

This PR is just enough for me to get started on splunk multisite testing. We need not take in this PR and wait for Ritesh's PR (I will build a binary on this branch for splunk testing)

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#6043a9a8-69f8-46ab-bf49-7716654507aa) |
<a name="6043a9a8-69f8-46ab-bf49-7716654507aa"></a>
### 10358-c4833cf/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:31717
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 34bb4e5eb8ee:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 6 minutes and 15 seconds
(4/15) Running aws-sdk-ruby tests ... done in 18 seconds
(5/15) Running awscli tests ... FAILED in 58 seconds
{
  "name": "awscli",
  "duration": 1519,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 3 minutes and 16 seconds
(7/15) Running minio-dotnet tests ... done in 1 minutes and 35 seconds
(8/15) Running minio-go tests ... done in 5 minutes and 42 seconds
(9/15) Running minio-java tests ... done in 6 minutes and 12 seconds
(10/15) Running minio-js tests ... done in 2 minutes and 39 seconds
(11/15) Running minio-py tests ... done in 10 minutes and 25 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 33 seconds
(13/15) Running s3select tests ... done in 50 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
Lets close this PR if this is not going to be merged and we are doing a more comprehensive change // cc @kerneltime

**Pull request closed without merging**